src/Development/IDE/Core/RuleTypes.hs

```diff
@@ -52,49 +54,68 @@ type instance RuleResult GetDependencies = TransitiveDependencies
 
 type instance RuleResult GetModuleGraph = DependencyInformation
 
+-- | Does this module need object code?
+type instance RuleResult NeedsObjectCode = Bool
+
 data GetKnownTargets = GetKnownTargets
   deriving (Show, Generic, Eq, Ord)
 instance Hashable GetKnownTargets
 instance NFData   GetKnownTargets
 instance Binary   GetKnownTargets
 type instance RuleResult GetKnownTargets = KnownTargets
 
+-- | Convert to Core, requires TypeCheck*
+type instance RuleResult GenerateCore = ModGuts
+
+data GenerateCore = GenerateCore
+    deriving (Eq, Show, Typeable, Generic)
+instance Hashable GenerateCore
+instance NFData   GenerateCore
+instance Binary   GenerateCore
+
+data GetImportMap = GetImportMap
+    deriving (Eq, Show, Typeable, Generic)
+instance Hashable GetImportMap
+instance NFData   GetImportMap
+instance Binary   GetImportMap
+
+type instance RuleResult GetImportMap = ImportMap
+newtype ImportMap = ImportMap
+  { importMap :: M.Map ModuleName NormalizedFilePath -- ^ Where are the modules imported by this file located?
+  } deriving stock Show
+    deriving newtype NFData
+
 -- | Contains the typechecked module and the OrigNameCache entry for
 -- that module.
 data TcModuleResult = TcModuleResult
-    { tmrModule     :: TypecheckedModule
```

Comment:
Is there any way to recover `tmrModule` from the new fields in `TcModuleResult`?  It's used by 3 plugins in HLS so recovering it would smooth things downstream.

If not, consider adding a helper in the `Rules` module and export it from `Development.IDE`:
```
 getTypecheckedModule :: NormalizedFilePath -> Action (Maybe TypecheckedModule)
```

Reply:
No, a `TypecheckedModule` contains `ModDetails`, which 

I don't think anything in hls needs `ModDetails`, which was not being generated properly for modules that used TH/need to be compiled, since it requires desugared source to properly generate (see my comment about `tidyProgram`) above.

The rest of the fields in a `TypecheckedModule` are available in `TcModuleResult`, so porting shouldn't be too much of an issue.